pkg/internal/managedappcredential/manager_ensure.go

```diff
@@ -0,0 +1,199 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedappcredential
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener-extension-provider-openstack/pkg/openstack"
+	openstackclient "github.com/gardener/gardener-extension-provider-openstack/pkg/openstack/client"
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+// Ensure will ensure the managed application credential of an Openstack Shoot cluster.
+func (m *Manager) Ensure(ctx context.Context, credentials *openstack.Credentials) error {
+	newParentUser, err := m.newParentFromCredentials(credentials)
+	if err != nil {
+		return err
+	}
+
+	appCredential, err := readApplicationCredential(ctx, m.client, m.namespace)
+	if err != nil {
+		return err
+	}
+
+	var (
+		appCredentialExists bool
+		parentChanged       bool
+		oldParentUserUsable bool
+
+		oldParentUser *parent
+	)
+
+	if appCredential != nil {
+		if err := m.updateParentPasswordIfRequired(ctx, appCredential, newParentUser); err != nil {
+			return err
+		}
+
+		if user, err := m.newParentFromSecret(appCredential.secret); err == nil && user != nil {
+			oldParentUser = user
+			oldParentUserUsable = true
+		}
+
+		if oldParentUserUsable {
+			if _, err := oldParentUser.identityClient.GetApplicationCredential(ctx, oldParentUser.id, appCredential.id); !openstackclient.IsNotFoundError(err) {
+				appCredentialExists = true
+			}
+
+			if oldParentUser.id != newParentUser.id {
+				parentChanged = true
+			}
+		}
+
+	}
+
+	if parentChanged {
+		// Try to clean up the application credentials owned by the old parent user.
+		// This might not work as the information about this user could be stale,
+		// because the user credentials are rotated, the user is not associated to
+		// Openstack project anymore or it is deleted.
+		if err := m.runGarbageCollection(ctx, oldParentUser, appCredential.id, false); err != nil {
+			m.logger.Error(err, "could not clean up application credential(s) as the owning user has changed and information about owning user might be stale")
+		}
+	}
+
+	// In case the application credential usage is disabled, it is tried to
+	// clean up old application credentials before aborting.
+	if !m.config.Enabled {
+		if !appCredentialExists {
+			return m.removeApplicationCredentialStore(ctx, appCredential.secret)
+		}
+
+		if oldParentUserUsable {
+			if err := m.runGarbageCollection(ctx, oldParentUser, "", true); err != nil {
+				return err
+			}
+		}
+
+		return m.removeApplicationCredentialStore(ctx, appCredential.secret)
+	}
+
+	// If the new parent user is itself an application credential, then no managed
+	// application credential can be used. It will be tried to clean up with the
+	// old user if this one is usuable.
+	if newParentUser.isApplicationCredential() {
+		if appCredentialExists {
```

Comment:
Do you need `oldParentUsable` check here too ? It seems similar to my previous comment so can they be consolidated e .g 

```
if !m.config.Enabled  || newParentUser.isApplicationCredential()
```

Reply:
Nice. Make sense.